planner/cascades/implementation_rules.go

```diff
@@ -221,3 +224,22 @@ func (r *ImplSort) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalP
 	)
 	return impl.NewSortImpl(ps), nil
 }
+
+// ImplHashAgg is the implementation rule which implements LogicalAggregation
+// to PhysicalHashAgg.
+type ImplHashAgg struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplHashAgg) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	// TODO: deal with the hints when we have implemented StreamAgg.
+	return prop.IsEmpty()
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplHashAgg) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	la := expr.ExprNode.(*plannercore.LogicalAggregation)
+	hashAgg := plannercore.NewPhysicalHashAgg(la, reqProp.ExpectedCnt, &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64})
+	// TODO: Implement TiKVHashAgg
+	return impl.NewTiDBHashAggImpl(hashAgg), nil
```

Comment:
Is there any problem to put the `engineType` into the `baseImpl`? We already have the `TableGather` now. If realizing numbers of engine type implementations for one operator. Isn't it more complex?

Reply:
We consider that different engine may have different behavior, not only costs.

For example, TiDBHashAgg needs Projection below the Agg if the AggFuncs' args contain ScalarFunction. But TiKVHashAgg does not need it. (In other words, TiKV supports `sum(a+b)`, but TiDB only supports `project a+b as c -> sum(c)`)If we build one Implementation for all of the engines, we might need `switch engine - case EngineTiDB:` in all of the methods like `SetCost` or `AttachChildren`. 

But if we use different Implementation for different engine, we only need to focus on one engine each implementation. And we only need to check the engine in the ImplementationRule. 

Anyway, we can change this way if we finally find out that put the `engineType` into the `baseImpl` is better.

Reply:
Okay.